code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm

```diff
@@ -0,0 +1,460 @@
+		TI.visible_message("<span class='xenowarning'>The space around [TI] abruptly stops shifting and wavering.</span>", \
+		"<span class='xenodanger'>We cease binding our essence to this place...</span>")
+		return fail_activate()
+
+	var/turf/T = get_turf(TI)
+	TI.visible_message("<span class='xenonotice'>\A shimmering point suddenly coalesces from the warped space around [T].</span>", \
+	"<span class='xenodanger'>We complete our work, binding our essence to this point.</span>", null, 5) //Fluff
+	var/obj/effect/xenomorph/warp_shadow/WB = new(T) //Create the new beacon.
+	playsound(T, 'sound/weapons/emitter.ogg', 25, 1)
+	QDEL_NULL(TI.warp_shadow) //Delete the old beacon
+	TI.warp_shadow = WB //Set our warp shadow
+	WB.dir = TI.dir //Have it imitate our facing
+	WB.pixel_x = TI.pixel_x //Inherit pixel offsets
+	teleport_debuff_aoe(WB) //SFX
+
+	succeed_activate()
+	add_cooldown()
+
+
+// ***************************************
+// *********** Hyperposition
+// ***************************************
+/datum/action/xeno_action/hyperposition
+	name = "Hyperposition"
+	action_icon_state = "hyperposition"
+	mechanics_text = "We teleport back to our warp shadow after a delay. The delay scales with the distance teleported."
+	plasma_cost = 25
+	cooldown_timer = 5 SECONDS
+	keybind_signal = COMSIG_XENOABILITY_HYPERPOSITION
+
+
+/datum/action/xeno_action/hyperposition/action_activate()
+
+	var/mob/living/carbon/xenomorph/wraith/TI = owner
+	var/obj/effect/xenomorph/warp_shadow/WB = TI.warp_shadow
+
+	if(!WB)
+		to_chat(TI, "<span class='xenodanger'>We have no warp shadow to teleport to!</span>")
+		return fail_activate()
+
+	var/area/A = get_area(WB)
+
+	TI.do_jitter_animation(2000) //Animation
+
+	var/distance = get_dist(TI, TI.warp_shadow) //Get the distance so we can calculate the wind up
+	var/hyperposition_windup = clamp(WRAITH_HYPERPOSITION_MIN_WINDUP, distance * 0.5 SECONDS - 5 SECONDS, WRAITH_HYPERPOSITION_MAX_WINDUP)
+
+	TI.visible_message("<span class='warning'>The air starts to violently roil and shimmer around [TI]!</span>", \
+	"<span class='xenodanger'>We begin to teleport to our warp shadow located at [A] (X: [WB.x], Y: [WB.y]). We estimate this will take [hyperposition_windup * 0.1] seconds.</span>")
+
+	if(!do_after(TI, hyperposition_windup, TRUE, TI, BUSY_ICON_FRIENDLY)) //Channel time/wind up
+		TI.visible_message("<span class='xenowarning'>The space around [TI] abruptly stops shifting and wavering.</span>", \
+		"<span class='xenodanger'>We abort teleporting back to our warp shadow.</span>")
+		return fail_activate()
+
+
+	hyperposition_warp() //We swap positions with our warp shadow
+
+	succeed_activate()
+	GLOB.round_statistics.wraith_hyperpositions++
+	SSblackbox.record_feedback("tally", "round_statistics", 1, "wraith_hyperpositions") //Statistics
+
+	add_cooldown()
+
+/datum/action/xeno_action/hyperposition/proc/hyperposition_warp() //This handles the location swap between the Wraith and the Warp Shadow
```

Comment:
These proc names are confusing this isnt a paper use proc names normal humans can read

Reply:
Also this proc should be inlined

Reply:
And isnt documented properly